src/main/java/org/apache/maven/plugins/clean/Cleaner.java

```diff
@@ -286,4 +393,189 @@ public void update( Result result )
+                        data.set( LAST_DIRECTORY_TO_DELETE, null );
+                        return lastDir;
+                    }
+                }
+                status = STOPPED;
+                notifyAll();
+            }
+            return basedir;
+        }
+
+        synchronized boolean doDelete( File dir )
+        {
+            if ( status == STOPPED )
+            {
+                return false;
+            }
+            filesToDelete.add( dir );
+            if ( status == NEW )
+            {
+                status = RUNNING;
+                notifyAll();
+                wrapExecutionListener();
+                start();
+            }
+            return true;
+        }
+
+        /**
+         * If this has not been done already, we wrap the ExecutionListener inside a proxy
+         * which simply delegates call to the previous listener.  When the session ends, it will
+         * also call {@link BackgroundCleaner#sessionEnd()}.
+         * There's no clean API to do that properly as this is a very unusual use case for a plugin
+         * to outlive its main execution.
+         */
+        private void wrapExecutionListener()
+        {
+            ExecutionListener executionListener = cleaner.session.getRequest().getExecutionListener();
+            if ( executionListener == null
+                    || !Proxy.isProxyClass( executionListener.getClass() )
+                    || !( Proxy.getInvocationHandler( executionListener ) instanceof SpyInvocationHandler ) )
+            {
+                ExecutionListener listener = ( ExecutionListener ) Proxy.newProxyInstance(
+                        ExecutionListener.class.getClassLoader(),
+                        new Class[] { ExecutionListener.class },
+                        new SpyInvocationHandler( executionListener ) );
+                cleaner.session.getRequest().setExecutionListener( listener );
+            }
+        }
+
+        synchronized void doSessionEnd()
+        {
+            if ( status != STOPPED )
+            {
+                try
+                {
+                    cleaner.logInfo.log( "Waiting for background file deletion" );
+                    while ( status != STOPPED )
+                    {
+                        wait();
+                    }
+                }
+                catch ( InterruptedException e )
+                {
+                    // ignore
+                }
+            }
+        }
+
+    }
+
+    static class SpyInvocationHandler implements InvocationHandler
+    {
+        private final ExecutionListener delegate;
+
+        SpyInvocationHandler( ExecutionListener delegate )
+        {
+            this.delegate = delegate;
+        }
+
+        @Override
+        public Object invoke( Object proxy, Method method, Object[] args ) throws Throwable
+        {
+            if ( "sessionEnded".equals( method.getName() ) )
+            {
+                BackgroundCleaner.sessionEnd();
```

Comment:
For my use case, it would be optimal if we could have e.g. a few different modes of when and how the delete takes place. Something like:
enum Mode {
  /* Start immediately deleting in fastmode, deleting takes place during build and completes at session end (blocks) */
  background,
  /* Deletes in fastmode at the end of the session (blocks) */
  atEnd,
  /* Start immediately deleting in fastmode, deleting takes place during build, ignores session end. Only applicable when running in a daemon, otherwise the delete might not complete */
  backgroundDaemon,
  /* Starts deleting in fastmode at session end. Only applicable when running in a daemon, otherwise the delete might not complete */
  atEndDaemon
}

Also, I have noticed that running the delete in multiple threads does indeed speed things up so perhaps have that option as well (another property).

I know the daemon mode doesn't make sense for mvn but it would mean the terminal would be freed up a lot sooner when running with mvn-daemon.

Reply:
Guess for mvnd case we only need to enable to override clean mojo easily and not add an useless option upfront so for me it is only atEnd and integrations can override it at need. I don't see mojo getting mvnd specific options because:

1. it would be useless option for all mvn builds so doc would be unlikely IMHO
2. there is no technical blocker to do what you want in mvnd as of today (worse is done to mention it)

I'm suspicious on the background cause it can create easily not deterministic behaviors if it is mixed with other commands watching target so can be better to use some checkpointing (synchronizations are more than sufficient) - but it is likely a daemon thing only.